### PR TITLE
Site Editor: Add navigation panel with placeholder content

### PR DESCRIPTION
**Pull request by david-szabo97**
Related: https://github.com/WordPress/gutenberg/issues/25260 https://github.com/WordPress/gutenberg/issues/25261 

## Description
Putting Navigation component in action! Great way to see how it cooperates with the site editor. Also acts an example of how we could implement the preview popup.

This is a minimal implementation to see the Navigation component in the Site Editor.

## Needs fix
- [x] Menu title text color is overridden by other CSS (https://github.com/WordPress/gutenberg/pull/25507)
- [x] Group title text color is overridden by other CSS (https://github.com/WordPress/gutenberg/pull/25507)

## How has this been tested?
* `yarn wp-env start`
* Enable FSE
* Open Site Editor
* Click site logo to toggle the navigation sidebar

## Screenshots
Closed:
image

Open:
image

Preview:
image

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/edit-site/src/components/header/index.js
Comment (vindl) [inline packages/edit-site/src/components/header/index.js:92]: This button is exported from the package (experimental for now) and provides a way for 3rd plugins to customize some of its parts while retaining functionality ([see docs here](https://github.com/WordPress/gutenberg/blob/65258a077e4899ee4933c79077108f56f8b5ce6d/docs/designers-developers/developers/slotfills/main-dashboard-button.md)). Given that it will no longer represent a close action, I think that we should rename the existing `FullscreenModeClose` to `NavigationButton ` or `DashboardButton` and place our code changes there.

File: packages/edit-site/src/components/editor/index.js
Comment (vindl) [inline packages/edit-site/src/components/editor/index.js:49]: Provided that we want this behavior, it might make sense to collapse these states into one `leftSidebarContent` that can have on of three possible values for now: `null`, `navigation`, and `inserter`. With present setup adding another element that wants to toggle into this space (there were some plans for doing that with block navigation) will be cumbersome, since each element will have to be aware of others and the toggling logic will have to be updated everywhere.

It would also remove the need for current naming with leading underscore (`setIsInserterOpen` and `_setIsInserterOpen`), which I find hard to read.

File: packages/edit-site/src/components/editor/index.js
Comment (vindl) [inline packages/edit-site/src/components/editor/index.js:177]: Should we wrap this block into helper function or make it a separate component?

**Review by vindl — COMMENTED** [review]
Nice work so far @david-szabo97! ✨

**Comment by david-szabo97** [comment]
#25507 is merged so I did a rebase. Updated the description with the new screenshots.

**Comment by vindl** [comment]
@shaunandrews I have a question about the intended behavior here. Opening the inserter will close the navigation sidebar and vice versa in present version. Is that fine or should it be possible for both sidebars to be open at the same time?

Also, how should it interact with potential [Persistent block navigator panel](https://github.com/WordPress/gutenberg/issues/22113)?

File: packages/edit-site/src/components/editor/index.js
Comment (david-szabo97) [reply packages/edit-site/src/components/editor/index.js:177]: Yep! I'd say go for a separate component. And if we are already there, we should extract inserter panel as a separate component as well. What do you think about this structure?

```
packages/edit-site/src/components/left-sidebar
packages/edit-site/src/components/left-sidebar/inserter-panel
packages/edit-site/src/components/left-sidebar/navigation-panel (note: NavigationPanel instead of NavigationSidebar)
```

File: packages/edit-site/src/components/editor/index.js
Comment (david-szabo97) [reply packages/edit-site/src/components/editor/index.js]: Done!

File: packages/edit-site/src/components/header/index.js
Comment (Copons) [reply packages/edit-site/src/components/header/index.js:92]: We should also figure out how to handle the FSE Navigation when not in fullscreen mode, where we wouldn't have the WP/Site Icon back button anymore.

File: packages/edit-site/src/components/editor/index.js
Comment (vindl) [reply packages/edit-site/src/components/editor/index.js:163]: Sounds fine. 👍

File: packages/edit-site/src/components/header/index.js
Comment (vindl) [reply packages/edit-site/src/components/header/index.js:92]: Good point @Copons! I don't consider it a blocker for this PR though, as it's something that we can figure out later.

**Comment by david-szabo97** [comment]
- Extracted left sidebar components (`LeftSidebar`, `InserterPanel`, `NavigationPanel`)
- Added a temporary NavigationItem to the `NavigationPanel` which points to the dashboard
- Renamed `NavigationButton` to `DashboardButton` and removed `FullscreenModeClose`

File: packages/edit-site/src/components/left-sidebar/index.js
Comment (vindl) [inline packages/edit-site/src/components/left-sidebar/index.js:8]: I don't think we need `leftSidebar` var anymore or `else if`. We simplify this to:

```javascript
if ( content === 'navigation' ) {
    return <NavigationPanel />;
}

if ( content === 'inserter' ) {
    return <InserterPanel closeInserter={ () => setContent( null ) } />;
}

return null;

```

Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/index.js]: Done! I was driven away by the original code 😓 😄

**Review by vindl — APPROVED** [review]
Thanks @david-szabo97! This looks like a good starting point for our upcoming iteratons.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:40]: How do you feel about removing all demo content except for "Back to dashboard" and "Hover to show preview"?
Everything else is just bloating the PR, and we can just look at the Storybook for usage examples.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:20]: No need for this fragment here.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:21]: What about adding a `.edit-site-navigation-panel` class here, and move these styles to `styles.scss`?

File: packages/edit-site/src/components/header/dashboard-button/style.scss
Comment (Copons) [inline packages/edit-site/src/components/header/dashboard-button/style.scss:1]: Please note that we should use a ["loose BEM" naming](https://github.com/WordPress/gutenberg/blob/f485f1f6a5556893b7774f2e9ef3320b0884c620/docs/contributors/coding-guidelines.md#css).
Modifier should be preceded by a double underscore.
E.g. `. edit-site-dashboard-button__wrapper`

(There are a few other occurrences of this in the PR, but I'll leave you the _fun_ of finding them 😛 )

File: packages/edit-site/src/components/header/index.js
Comment (Copons) [inline packages/edit-site/src/components/header/index.js:97]: I've got a few doubts about this change.
While it's true that this button doesn't close fullscreen mode anymore, it also doesn't navigate back to the dashboard (so the "Open dashboard" label in the component is misleading) — and it's not even technically _just_ a button 😄.

What it does now is to toggle the navigation sidebar, so I'd be more inclined in calling it `NavigationPanelToggle`.

---

I think we could also keep `FullscreenModeClose`.
While it would be left unused in this PR, we could update it to become (_be_?) the actual "dashboard button".
Consumers relying on it would not break, and we could use it inside the navigation toggle, or in the navigation itself (or wherever we see fit).

Either way, my opinion is that we should probably avoid removing existing components from an experimental PR whose purpose is to be a starting point, and not the finished product.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:16]: The `activeMenu` state is optional, only useful if we need to control the hierarchical navigation from outside.
It's not really needed here at this point.

File: packages/edit-site/src/components/editor/style.scss
Comment (Copons) [inline packages/edit-site/src/components/editor/style.scss:31]: This is probably not the right place for these styles.
What about:
- Remove everything from here.
- Create the `.edit-site-navigation-panel` class as suggested in [another comment](https://github.com/WordPress/gutenberg/pull/25506/files#r493329611).
- Add something like this in `packages/edit-site/src/components/left-sidebar/navigation-panel/style.scss`? 
```scss
.edit-site-navigation-panel {
	height: 100%;
	position: relative;
	width: 300px;

	.components-navigation {
		height: 100%;
	}
}
```

File: packages/edit-site/src/components/header/index.js
Comment (vindl) [reply packages/edit-site/src/components/header/index.js:97]: I agree that that name is more appropriate. Also no objections to keeping the `FullscreenModeClose` if we are going to repurpose it and include it in navigation.

Comment (Copons) [reply packages/edit-site/src/components/header/index.js:97]: My point is also that the navigation sidebar is still in flux, and I'm not super keen on removing an established component (albeit experimental).
We might end up not using it, or heavily modify it, but let's care about that when we'll be actually working on the back navigation. 🙂

Comment (david-szabo97) [reply packages/edit-site/src/components/header/index.js:97]: - Renamed `DashboardButton` to `NavigationToggle`
- Reverted `FullscreenModeClose` component, let's keep it for now and we will iterate upon it

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js
Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/index.js]: Removed them but I left a `NavigationGroup` just in case. Once this merged I'll put some real data in the navigation sidebar so this phase is kinda temporarily.

File: packages/edit-site/src/components/header/dashboard-button/style.scss
Comment (david-szabo97) [reply packages/edit-site/src/components/header/dashboard-button/style.scss]: I was confused by `FullscreenModeClose` because that component is not following the BEM. 🤔 Thanks for letting me know, fixed it up!

**Review by Copons — APPROVED** [review]
This LGTM!
Thank you @david-szabo97 for working on this! ✨ 

I'll open an issue for future discussion of the `FullscreenModeClose` destiny.

**Review by mattwiebe — APPROVED** [review]
This looks good to me. Left one piece of feedback that we can address in a renaming PR but nothing to block this.

**Comment by afercia** [comment]
First observation: the `W` button was recently changed to accept the Site Icon set in the Customizer as a replacement for hte "W". In the block editor, a site icon looks like this:

Screenshot 2020-09-24 at 18 02 27

In the Site Editor, it looks like this 🙂 : 

Screenshot 2020-09-24 at 18 02 05

where apparently there's some missing CSS.

Besides that, I'm not sure using the same UI control for different purposes on different screens is a good user experience. As a user, If I know the W button in the block editor brings me back to the Posts lists, I would expect the same button in the Site editor to do the same.

**Comment by afercia** [comment]
Worth also noting the "sliding in sidebar" pattern is undergoing discussion by the accessibility team as it's a pattern that introduces various accessibility issues. It does relate to other new sidebars as well though so it's not specific to this PR. An audit is going to be made on the issue related to the main Inserter, that was already identified by the accessibility team as an accessibility regression in WordPress 5.5 compared to the implementation in 5.4. See #24975

**Comment by Copons** [comment]
@afercia Thanks for chiming in!
The site icon size bug will be fixed in #25622.

The plan here was to merge the sidebar in the Site Editor in order to work on it in a real scenario, but still "protected" by the fact that it's an experimental environment.

We have been rushing this PR in because we (as in: people working on the Site Editor) are trying to improve the whole UX, which is arguably not very good right now.
One of the steps is to move the template switcher out of the header toolbar and into this sidebar. This will unblock other unrelated work that needs more room in the header toolbar, while keeping the template switching functionality available.

Thank you for linking that accessibility issue.
Admittedly, I haven't had time to look into it (I'm not an a11y expert, so it will take some time 🙂), but I promise we'll get to it.

**Pull request merged**
